File: generated-code/typescript/codecraft/FileReadWrite/src/codegame/message-game-model.ts
import { ClientMessage } from "./client-message";
import { ServerMessage } from "./server-message";
import { Stream } from "../stream";

/**
 * Client or server message
 */
export abstract class MessageGameModel {
    /**
     * Write MessageGameModel to output stream
     */
    abstract writeTo(stream: Stream): Promise<void>;

    /**
     * Read MessageGameModel from input stream
     */
    static async readFrom(stream: Stream): Promise<MessageGameModel> {
        const tag = await stream.readInt();
        if (tag == MessageGameModel.Client.TAG) {
            return await MessageGameModel.Client.readFrom(stream);
        }
        if (tag == MessageGameModel.Server.TAG) {
            return await MessageGameModel.Server.readFrom(stream);
        }
        throw new Error("Unexpected tag value");
    }
}

export namespace MessageGameModel {
    /**
     * Client message
     */
    export class Client extends MessageGameModel {
        /**
         * Message
         */
        message: ClientMessage
    
        constructor(message: ClientMessage) {
            super();
            this.message = message;
        }
    
        /**
         * Read Client from input stream
         */
        static async readFrom(stream: Stream): Promise<MessageGameModel.Client> {
            let message;
            message = await ClientMessage.readFrom(stream);
            return new Client(message)
        }
    
        /**
         * Write Client to output stream
         */
        async writeTo(stream: Stream) {
            await stream.writeInt(Client.TAG);
            let message = this.message;
            await message.writeTo(stream);
        }
    }
    
    export namespace Client {
        export const TAG = 0;
    }
    /**
     * Server message
     */
    export class Server extends MessageGameModel {
        /**
         * Message
         */
        message: ServerMessage
    
        constructor(message: ServerMessage) {
            super();
            this.message = message;
        }
    
        /**
         * Read Server from input stream
         */
        static async readFrom(stream: Stream): Promise<MessageGameModel.Server> {
            let message;
            message = await ServerMessage.readFrom(stream);
            return new Server(message)
        }
    
        /**
         * Write Server to output stream
         */
        async writeTo(stream: Stream) {
            await stream.writeInt(Server.TAG);
            let message = this.message;
            await message.writeTo(stream);
        }
    }
    
    export namespace Server {
        export const TAG = 1;
    }
}

File: generated-code/ruby/codecraft/FileReadWrite/codegame/message_game_model.rb
require './codegame/client_message'
require './codegame/server_message'

module Codegame

# Client or server message
class MessageGameModel
    # Read MessageGameModel from input stream
    def self.read_from(stream)
        tag = stream.read_int()
        if tag == MessageGameModel::Client::TAG
            return MessageGameModel::Client.read_from(stream)
        end
        if tag == MessageGameModel::Server::TAG
            return MessageGameModel::Server.read_from(stream)
        end
        raise "Unexpected tag value"
    end

    # Client message
    class Client
        TAG = 0
    
        # Message
        attr_accessor :message
    
        def initialize(message)
            @message = message
        end
    
        # Read Client from input stream
        def self.read_from(stream)
            message = Codegame::ClientMessage.read_from(stream)
            Client.new(message)
        end
    
        # Write Client to output stream
        def write_to(stream)
            stream.write_int(TAG)
            @message.write_to(stream)
        end
    
        def to_s
            string_result = "Client { "
            string_result += "message: "
            string_result += @message.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
    # Server message
    class Server
        TAG = 1
    
        # Message
        attr_accessor :message
    
        def initialize(message)
            @message = message
        end
    
        # Read Server from input stream
        def self.read_from(stream)
            message = Codegame::ServerMessage.read_from(stream)
            Server.new(message)
        end
    
        # Write Server to output stream
        def write_to(stream)
            stream.write_int(TAG)
            @message.write_to(stream)
        end
    
        def to_s
            string_result = "Server { "
            string_result += "message: "
            string_result += @message.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
end

end